$OpenBSD: patch-tools_clang_lib_Basic_Targets_cpp,v 1.19 2016/12/22 11:07:13 ajacoutot Exp $

- Enable __float128 support on X86
- [OpenBSD] Add type sign information for OpenBSD
- Set ABI for ARM
- [OpenBSD] set OpenBSD defines for aarch64

--- tools/clang/lib/Basic/Targets.cpp.orig	Tue Nov 29 10:53:17 2016
+++ tools/clang/lib/Basic/Targets.cpp	Wed Dec 21 19:39:24 2016
@@ -542,6 +542,8 @@ class OpenBSDTargetInfo : public OSTargetInfo<Target> 
     Builder.defineMacro("__ELF__");
     if (Opts.POSIXThreads)
       Builder.defineMacro("_REENTRANT");
+    if (this->HasFloat128)
+      Builder.defineMacro("__FLOAT128__");
   }
 public:
   OpenBSDTargetInfo(const llvm::Triple &Triple, const TargetOptions &Opts)
@@ -550,8 +552,11 @@ class OpenBSDTargetInfo : public OSTargetInfo<Target> 
 
       switch (Triple.getArch()) {
         default:
+          break;
         case llvm::Triple::x86:
         case llvm::Triple::x86_64:
+          this->HasFloat128 = true;
+          // FALLTHROUGH
         case llvm::Triple::arm:
         case llvm::Triple::sparc:
           this->MCountName = "__mcount";
@@ -4641,8 +4646,10 @@ class ARMTargetInfo : public TargetInfo {
     DoubleAlign = LongLongAlign = LongDoubleAlign = SuitableAlign = 64;
     const llvm::Triple &T = getTriple();
 
-    // size_t is unsigned long on MachO-derived environments, NetBSD and Bitrig.
+    // size_t is unsigned long on MachO-derived environments, NetBSD,
+    // OpenBSD and Bitrig.
     if (T.isOSBinFormatMachO() || T.getOS() == llvm::Triple::NetBSD ||
+        T.getOS() == llvm::Triple::OpenBSD ||
         T.getOS() == llvm::Triple::Bitrig)
       SizeType = UnsignedLong;
     else
@@ -4650,6 +4657,7 @@ class ARMTargetInfo : public TargetInfo {
 
     switch (T.getOS()) {
     case llvm::Triple::NetBSD:
+    case llvm::Triple::OpenBSD:
       WCharType = SignedInt;
       break;
     case llvm::Triple::Win32:
@@ -4849,6 +4857,7 @@ class ARMTargetInfo : public TargetInfo {
 
     switch (getTriple().getOS()) {
     case llvm::Triple::NetBSD:
+    case llvm::Triple::OpenBSD:
       PtrDiffType = SignedLong;
       break;
     default:
@@ -4901,6 +4910,8 @@ class ARMTargetInfo : public TargetInfo {
       default:
         if (Triple.getOS() == llvm::Triple::NetBSD)
           setABI("apcs-gnu");
+        else if (Triple.getOS() == llvm::Triple::OpenBSD)
+          setABI("aapcs-linux");
         else
           setABI("aapcs");
         break;
@@ -8161,6 +8172,8 @@ static TargetInfo *AllocateTarget(const llvm::Triple &
       return new LinuxTargetInfo<AArch64leTargetInfo>(Triple, Opts);
     case llvm::Triple::NetBSD:
       return new NetBSDTargetInfo<AArch64leTargetInfo>(Triple, Opts);
+    case llvm::Triple::OpenBSD:
+      return new OpenBSDTargetInfo<AArch64leTargetInfo>(Triple, Opts);
     default:
       return new AArch64leTargetInfo(Triple, Opts);
     }
